tests/kubernetes-e2e.rs

```diff
@@ -0,0 +1,732 @@
+            // one.
+            panic!("marker seen more than once");
+        }
+
+        // If we did, remember it.
+        got_marker = true;
+
+        // Request to stop the flow.
+        FlowControlCommand::Terminate
+    })
+    .await?;
+
+    assert!(got_marker);
+
+    drop(test_pod);
+    drop(test_namespace);
+    drop(vector);
+    Ok(())
+}
+
+/// This test validates that vector picks up multiple log lines, and that they
+/// arrive at the proper order.
+#[tokio::test]
+async fn multiple_lines() -> Result<(), Box<dyn std::error::Error>> {
+    let _guard = lock();
+    let framework = make_framework();
+
+    let vector = framework.vector("test-vector", VECTOR_CONFIG).await?;
+    framework
+        .wait_for_rollout("test-vector", "daemonset/vector", vec!["--timeout=30s"])
+        .await?;
+
+    let test_namespace = framework.namespace("test-vector-test-pod").await?;
+
+    let test_messages = vec!["MARKER1", "MARKER2", "MARKER3", "MARKER4", "MARKER5"];
+    let test_pod = framework
+        .test_pod(test_pod::Config::from_pod(&make_test_pod(
+            "test-vector-test-pod",
+            "test-pod",
+            &format!("echo -e {}", test_messages.join(r"\\n")),
+            vec![],
+        ))?)
+        .await?;
+    framework
+        .wait(
+            "test-vector-test-pod",
+            vec!["pods/test-pod"],
+            WaitFor::Condition("initialized"),
+            vec!["--timeout=30s"],
+        )
+        .await?;
+
+    let mut log_reader = framework.logs("test-vector", "daemonset/vector")?;
+    smoke_check_first_line(&mut log_reader).await;
+
+    // Read the rest of the log lines.
+    let mut test_messages_iter = test_messages.into_iter().peekable();
+    look_for_log_line(&mut log_reader, |val| {
+        if val["kubernetes"]["pod_namespace"] != "test-vector-test-pod" {
+            // A log from something other than our test pod, pretend we don't
+            // see it.
+            return FlowControlCommand::GoOn;
+        }
+
+        // Take the next marker.
+        let current_marker = test_messages_iter
+            .next()
+            .expect("expected no more lines since the test messages iter is exausted");
+
+        // Ensure we got the marker.
+        assert_eq!(val["message"], current_marker);
+
+        if test_messages_iter.peek().is_some() {
+            // We're not done yet, so go on.
+            return FlowControlCommand::GoOn;
+        }
+
+        // Request to stop the flow.
+        FlowControlCommand::Terminate
+    })
+    .await?;
+
+    assert!(test_messages_iter.next().is_none());
+
+    drop(test_pod);
+    drop(test_namespace);
+    drop(vector);
+    Ok(())
+}
+
+/// This test validates that vector properly annotates log events with pod
+/// metadata obtained from the k8s API.
+#[tokio::test]
+async fn pod_metadata_annotation() -> Result<(), Box<dyn std::error::Error>> {
+    let _guard = lock();
```

Comment:
Since this is isolated to a cargo run you could use `RUST_TEST_THREADS` to force the runner to only run 1 test at a time here. That'd avoid the complexity of the mutex maybe?

Reply:
This one is sort of a primary safety measure, while `RUST_TEST_THREADS` is secondary. This is to avoid a situation where running without `RUST_TEST_THREADS` causes test failures. It makes more sense to keep such restriction as an explicit part of the test.

At the other hand, `RUST_TEST_THREADS` makes test improves output readability - when tests run in parallel and lock on the mutex, they work alright, but the start and the relevant output of the test can go separately - as multiple tests print their starting lines at the same time, and then they run in sequence - very confusing.

Reply:
Yeah it's kind of a toss up. I think having both is a sane solution that measures the costs of both. :)